File: CTCI/Ch2LinkedLists/Py/duplicates.py
"""
This algorithm removes dupliactes
from a Linked List
"""


class Node(object):
    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next_node = next_node


class LinkedList:
    head = None

    @property
    def print(self):
        temp = self.head
        while temp is not None:
            print(temp.data + '-->', end='')
            temp = temp.next_node

    def insert(self, data):
        new_node = Node(data, None)
        if self.head is None:
            self.head = new_node
        new_node.next_node = self.head
        self.head = new_node

    def remove(self, data):
        temp = self.head
        while temp is not None:
            if temp.data == data:
                temp.next_node = temp.next_node.next_node
            else:
                temp = temp.next_node

    def duplicatesWithBuf(self):
        temp = self.head
        vals = []
        while temp.next is not None:
            if temp.data in vals:
                temp = temp.next_node
            else:
                vals.append(temp.data)
                temp = temp.next_node
        for val in vals:
            print(val + '-->')

    def duplicatesWithoutBuf(self):
        current = self.head
        while current is not None:
            runner = self.head
            while runner is not None:
                if runner.data != current.data:
                    runner.next_node = runner.next_node.next_node
                else:
                    runner = runner.next_node

File: CTCI/Ch2LinkedLists/java/Duplicates.java
import java.util.*;
/**
  * This class removes the duplicates from a 
  * Linked List
  * Examples:
  * 3-->3-->4-->5-->1-->END => 4-->5-->1-->END
  */

public class Duplicates {
	protected Node head;
	
	public class Node {
		int data;
		Node next;

		public Node() {
			next = null;
		}

		public Node(int nodeData) {
			data = nodeData;
		}
	}
	public Duplicates() {
		head = null;
	}

	public void addFirst(int data) {
		if (head == null) {
			head = new Node(data);
		} else {
			Node temp = new Node(data);
			temp.next = head;
			head = temp;
		}
	}

	public void print(Node head) {
		Node temp = head;
		while (temp.next != null) {
			System.out.print(temp.data + "-->");
			temp = temp.next;
		}
	}

	public void removeDuplicates(Node head) {
		Set<Integer> set = new HashSet<Integer>();
		Node temp = head;
		while(temp.next != null) {
			set.add(temp.data);
			temp = temp.next;
		}
		Iterator iter = set.iterator();
		while(iter.hasNext()) {
			System.out.print(iter.next() + "-->");
		}
	}

	public static void main(String[] args) {
		Duplicates d = new Duplicates();
		d.addFirst(1);
		d.addFirst(3);
		d.addFirst(3);
		d.addFirst(5);
		d.addFirst(6);
		d.addFirst(7);
		d.addFirst(5);
		System.out.println("List before dup removal");
		d.print(d.head);
		System.out.println("\nList after dup removal");
		d.removeDuplicates(d.head);
	}
}
